### PR TITLE
Introduce APIs for fee payers

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

Introduce APIs for fee payers

RPC test for all tx types will be added on the RPC tester after this PR merged.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments
**RPC Test Result**
SignTransactionAsFeePayer
```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_signTransactionAsFeePayer", "params": [{"typeInt": 9, "from": "0x0fc6bc72c6481bb44660cf8a572f831cc8c35c95", "to": "0x44711E89b0c23845b5B2ed9D3716BA42b8a3e075", "gas": "0x76c0", "gasPrice": "0x5d21dba00", "value": "0xf4", "feePayer": "0x819104a190255e0cedbdd9d5f59a557633d79db1"}], "id": 92}' http://127.0.0.1:8551
{
    "id": 92,
    "jsonrpc": "2.0",
    "result": {
        "raw": "0x09f8953c8505d21dba008276c09444711e89b0c23845b5b2ed9d3716ba42b8a3e07581f4940fc6bc72c6481bb44660cf8a572f831cc8c35c95c094819104a190255e0cedbdd9d5f59a557633d79db1f847f845824e43a046e1abf3a13af79a94d9544b04ce25fa6f1922a0fb522fc2ceef0c51a5cccc72a0584a93dfada91387e47d402c774a4d7b63dafa2f505629240f38f0e7a3e35616",
        "tx": {
            "feePayer": "0x819104a190255e0cedbdd9d5f59a557633d79db1",
            "feePayerSignatures": [
                {
                    "R": "0x46e1abf3a13af79a94d9544b04ce25fa6f1922a0fb522fc2ceef0c51a5cccc72",
                    "S": "0x584a93dfada91387e47d402c774a4d7b63dafa2f505629240f38f0e7a3e35616",
                    "V": "0x4e43"
                }
            ],
            "from": "0x0fc6bc72c6481bb44660cf8a572f831cc8c35c95",
            "gas": "0x76c0",
            "gasPrice": "0x5d21dba00",
            "hash": "0xd5afb40a7a65b3041be94e491dd4da970bcc8dc17efd7c6cdb09418165023368",
            "nonce": "0x3c",
            "signatures": [],
            "to": "0x44711e89b0c23845b5b2ed9d3716ba42b8a3e075",
            "type": "TxTypeFeeDelegatedValueTransfer",
            "typeInt": 9,
            "value": "0xf4"
        }
    }
}
```
SendTransactionAsFeePayer
```
curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_sendTransactionAsFeePayer", "params": [{"typeInt": 9, "from": "0x0fc6bc72c6481bb44660cf8a572f831cc8c35c95", "to": "0x44711E89b0c23845b5B2ed9D3716BA42b8a3e075", "gas": "0x2e6300", "gasPrice": "0x5d21dba00", "value": "0xf4", "feePayer": "0x819104a190255e0cedbdd9d5f59a557633d79db1", "signatures": [{"V": "0x4e44", "R": "0x1692a48f166e3ef146eba61cbd6b450926854bb340cf6f689239f27588159419", "S": "0x277b9c6e97dfa6fdbcc15c201b5f7e05a2ef03f6247d9b9c541f98b8c4f1041a"}]}], "id": 57}' http://127.0.0.1:8551
{
    "id": 57,
    "jsonrpc": "2.0",
    "result": "0x3165e273fbea83bf90169bdf339532fe8515e09aee0f86ae5732ab2ab498af15"
}

curl -H "Content-Type: application/json" --data '{"jsonrpc": "2.0", "method": "klay_getTransactionReceipt", "params": ["0x3165e273fbea83bf90169bdf339532fe8515e09aee0f86ae5732ab2ab498af15"], "id": 99}' http://127.0.0.1:8551
{
    "id": 99,
    "jsonrpc": "2.0",
    "result": {
        "blockHash": "0x1ef101f21509e66a7d6b6523c1e5590db01e75a8c086e99f3c7eb2ac817f77c4",
        "blockNumber": "0x9a92",
        "contractAddress": null,
        "feePayer": "0x819104a190255e0cedbdd9d5f59a557633d79db1",
        "feePayerSignatures": [
            {
                "R": "0xb65745fa192505855a6ea3df4bdcdb781943a3db3b4a0c37add0db196b5f7c02",
                "S": "0x4075f5a16f918e9f8f5f575377adad415aaacf5accf94e153032a61e7bd539bb",
                "V": "0x4e43"
            }
        ],
        "from": "0x0fc6bc72c6481bb44660cf8a572f831cc8c35c95",
        "gas": "0x2e6300",
        "gasPrice": "0x5d21dba00",
        "gasUsed": "0x7918",
        "logs": [],
        "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "nonce": "0x3b",
        "senderTxHash": "0x4bb7ad0305a9d32634f548c65023aa3e7433f0b221c0eae1bcead1e074d455e2",
        "signatures": [
            {
                "R": "0x1692a48f166e3ef146eba61cbd6b450926854bb340cf6f689239f27588159419",
                "S": "0x277b9c6e97dfa6fdbcc15c201b5f7e05a2ef03f6247d9b9c541f98b8c4f1041a",
                "V": "0x4e44"
            }
        ],
        "status": "0x1",
        "to": "0x44711e89b0c23845b5b2ed9d3716ba42b8a3e075",
        "transactionHash": "0x3165e273fbea83bf90169bdf339532fe8515e09aee0f86ae5732ab2ab498af15",
        "transactionIndex": "0x0",
        "type": "TxTypeFeeDelegatedValueTransfer",
        "typeInt": 9,
        "value": "0xf4"
    }
}
```

### CODE REVIEW

File: api/api_private_account.go [hunk 1]
```diff
@@ -246,6 +246,29 @@ func (s *PrivateAccountAPI) SendTransaction(ctx context.Context, args SendTxArgs
 	return submitTransaction(ctx, s.b, signedTx.Tx)
 }
 
+// SendTransactionAsFeePayer will create a transaction from the given arguments and
+// tries to sign it as a fee payer with the key associated with args.From. If the
+// given passwd isn't able to decrypt the key it fails.
+func (s *PrivateAccountAPI) SendTransactionAsFeePayer(ctx context.Context, args SendTxArgs, passwd string) (common.Hash, error) {
+	if args.Nonce == nil {
```
Comment (KimKyungup) [inline api/api_private_account.go]: Is this nonce the sender's one?
I think this nonce is not related with feePayer. So we don't need to lock/unlock it.
But Shouldn't the fee payer sign the transaction with nil nonce?

File: api/api_private_account.go [hunk 2]
```diff
@@ -334,6 +357,38 @@ func (s *PrivateAccountAPI) SignTransaction(ctx context.Context, args SendTxArgs
 	return &SignTransactionResult{data, signedTx}, nil
 }
 
+// SignTransactionAsFeePayer will create a transaction from the given arguments and
+// tries to sign it as a fee payer with the key associated with args.From. If the given
+// passwd isn't able to decrypt the key, it fails. The transaction is returned in RLP-form,
+// not broadcast to other nodes
+func (s *PrivateAccountAPI) SignTransactionAsFeePayer(ctx context.Context, args SendTxArgs, passwd string) (*SignTransactionResult, error) {
+	// No need to obtain the noncelock mutex, since we won't be sending this
```
Comment (KimKyungup) [inline api/api_private_account.go]: I think we don't need to say about fee payer's nonce lock.

File: api/api_private_account.go (hunk 1 above)
Comment (KimKyungup) [reply api/api_private_account.go]: Do you assume the sender account and the fee payer account in the same node?

Comment (KimKyungup) [reply api/api_private_account.go]: If then why don't you make a one sign API to sign both signature once?

File: api/api_public_transaction_pool_test.go [hunk 3]
```diff
@@ -128,25 +138,35 @@ func TestTxTypeSupport(t *testing.T) {
...
 	}
 }
 
 // testTxTypeSupport_normalCase test APIs with proper SendTxArgs values.
 func testTxTypeSupport_normalCase(t *testing.T, api PublicTransactionPoolAPI, ctx context.Context, args SendTxArgs) {
-	// test tx type support of SignTransaction
-	{
-		_, err := api.SignTransaction(ctx, args)
+	var err error
+	// TODO - more test cases will be added soon
```
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply api/api_public_transaction_pool_test.go]: I changed them. Thank you.

File: api/api_private_account.go (hunk 1 above)
Comment (aidan-kwon) [reply api/api_private_account.go]: You are right. The nonce lock is not required. Thank you!

File: api/api_private_account.go (hunk 2 above)
Comment (aidan-kwon) [reply api/api_private_account.go]: I changed the comment. Please check it again.

File: api/api_public_transaction_pool.go [hunk 4]
```diff
@@ -323,6 +336,26 @@ func (s *PublicTransactionPoolAPI) SendTransaction(ctx context.Context, args Sen
 	return submitTransaction(ctx, s.b, signedTx.Tx)
 }
 
+// SendTransactionAsFeePayer creates a transaction for the given argument, sign it as a fee payer
+// and submit it to the transaction pool.
+func (s *PublicTransactionPoolAPI) SendTransactionAsFeePayer(ctx context.Context, args SendTxArgs) (common.Hash, error) {
+	// Don't allow dynamic assign of the nonce value since the sender already signed on a specific nonce value.
+	if args.Nonce == nil {
```
Comment (KimKyungup) [inline api/api_public_transaction_pool.go]: Don't we need to check gasPrice and gasLimit also?

Comment (aidan-kwon) [reply api/api_public_transaction_pool.go]: Thank you. I changed it to return errors for the values set by `setDefaults`

**Comment by aidan-kwon** [comment]
Reviewers, please take a look. I rebased it to resolve a conflict.

File: api/api_private_account.go [hunk 5]
```diff
@@ -334,6 +364,38 @@ func (s *PrivateAccountAPI) SignTransaction(ctx context.Context, args SendTxArgs
 	return &SignTransactionResult{data, signedTx}, nil
 }
 
+// SignTransactionAsFeePayer will create a transaction from the given arguments and
+// tries to sign it as a fee payer with the key associated with args.From. If the given
```
Comment (ehnuje) [inline api/api_private_account.go]:
```suggestion
// try to sign it as a fee payer with the key associated with args.From. If the given
```

File: api/api_private_account.go [hunk 6]
```diff
@@ -334,6 +364,38 @@ func (s *PrivateAccountAPI) SignTransaction(ctx context.Context, args SendTxArgs
...
+// passwd isn't able to decrypt the key, it fails. The transaction is returned in RLP-form,
```
Comment (ehnuje) [inline api/api_private_account.go]:
```suggestion
// password isn't able to decrypt the key, it fails. The transaction is returned in RLP-form,
```

File: api/api_private_account.go [hunk 7]
```diff
@@ -360,6 +422,20 @@ func (s *PrivateAccountAPI) sign(addr common.Address, passwd string, tx *types.T
 	return wallet.SignTxWithPassphrase(account, passwd, tx, s.b.ChainConfig().ChainID)
 }
 
+// signAsFeePayer is a helper function that signs a transaction with the private key of the given address.
+// If the given passwd isn't able to decrypt the key, it fails.
```
Comment (ehnuje) [inline api/api_private_account.go]:
```suggestion
// If the given password isn't able to decrypt the key, it fails.
```

**Comment by aidan-kwon** [comment]
@andybclee Please, take a look.
